### PR TITLE
buildGoModule: announce removal of `buildFlags` and `buildFlagsArray`

### PR DESCRIPTION
**Pull request by katexochen**
## Description of changes

This has been deprecated for 2 years now, and removing it will simplify the builder in some places.

So I would suggest we announce the removal now so that

- The removal notice will be part of 24.05 release
- We can remove `buildFlags`/`buildFlagsArray` in the 24.11 cycle

---

Tree-wide cleanup: I've searched for remaining occurrences with

```
rg -U --multiline-dotall -l 'buildGoModule.*buildFlags'  
```

and update them to a more conventional usage. These were mostly transferred to `checkFlags`, some to using `ldflags`. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by katexochen** [comment]
I have tested the tree-wide cleanup by cherry-picking it on top of master and running `nixpkgs-review rev HEAD`:

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>31 packages built:</summary>
  <ul>
    <li>adb-sync</li>
    <li>blockbook</li>
    <li>corerad</li>
    <li>fishPlugins.wakatime-fish</li>
    <li>gcsfuse</li>
    <li>git-town</li>
    <li>gitbatch</li>
    <li>go-mtpfs</li>
    <li>goconvey</li>
    <li>google-cloud-bigtable-tool</li>
    <li>google-cloud-sql-proxy</li>
    <li>gucci</li>
    <li>ignite</li>
    <li>incus</li>
    <li>incus-unwrapped</li>
    <li>infracost</li>
    <li>lnd</li>
    <li>lxd</li>
    <li>lxd-unwrapped</li>
    <li>massren</li>
    <li>podman-tui</li>
    <li>prometheus-graphite-exporter</li>
    <li>prometheus-mysqld-exporter</li>
    <li>pulsarctl</li>
    <li>pulumi</li>
    <li>skeema</li>
    <li>sunpaper</li>
    <li>traefik</li>
    <li>trickster</li>
    <li>wakatime</li>
    <li>wallutils</li>
  </ul>
</details>

I've also ensured the ldflags injection still works correct for the affected packages:

```
$ result/bin/traefik version
Version:      2.10.7
Codename:     cheddar
Go version:   go1.21.5
Built:        I don't remember exactly
OS/Arch:      linux/amd64

$ result/bin/pulsarctl --version
Release Version: 2.11.1.3
Git Commit Hash: v2.11.1.3
Git Branch: None
UTC Build Time: None
Go Version: go1.21.5

$ result/bin/corerad --version
flag provided but not defined: -version
CoreRAD v1.2.2 (2023-04-25)
```

File: pkgs/applications/blockchains/lnd/default.nix
Comment (katexochen) [inline pkgs/applications/blockchains/lnd/default.nix:25]: These ldflags are not required, upstream read these values from runtime, see https://github.com/lightningnetwork/lnd/blob/2fb150c8fe827df9df0520ef9916b3afb7b03a8d/build/version.go#L66-L75

**Comment by katexochen** [comment]
I thought about the test skipping pattern a bit, 

```nix
"-skip=${builtins.concatStringsSep "|" skippedTests}"
```

maybe a more precise and still simple pattern would look like this

```nix
"-skip=^${builtins.concatStringsSep "$|^" skippedTests}$"
```
This will ensure exact matches while also working when excluding sub tests.

Somewhere I have spotted 

```nix
"-skip=^(${builtins.concatStringsSep "|" skippedTests})$"
```

which won't work when excluding sub tests (`TestA/SubTestB`) because of the slash handling (I guess).

**Comment by katexochen** [comment]
Rebased on latest staging to resolve merge conflicts.

**Review by kalbasit — APPROVED** [review]

**Review by RaghavSood — APPROVED** [review]

**Comment by katexochen** [comment]
Rebased to resolve merge conflict in `incus-unwrapped`.

**Comment by kalbasit** [comment]
@katexochen what are we waiting for to merge?

**Comment by katexochen** [comment]
> what are we waiting for to merge?

@kalbasit ready to merge from my side, but I don't have the commit bit (yet). :wink:

**Comment by katexochen** [comment]
@zowoq @kalbasit any chance you can merge this before the 24.05 freeze for breaking changes?

File: pkgs/by-name/lx/lxd-unwrapped-lts/package.nix
Comment (adamcstephens) [inline pkgs/by-name/lx/lxd-unwrapped-lts/package.nix:17]: Please undo the formatting changes.

**Comment by adamcstephens** [comment]
@ofborg build incus lxd-unwrapped

File: pkgs/by-name/lx/lxd-unwrapped-lts/package.nix
Comment (katexochen) [reply pkgs/by-name/lx/lxd-unwrapped-lts/package.nix]: sorry, fixed.

**Review by adamcstephens — APPROVED** [review]
Looks good for lxd/incus.

**Pull request merged**
